openmmmcmc/states.py

```diff
@@ -0,0 +1,1903 @@
+        """Check if the system is consistent with the state.
+
+        It raises an Exception if the system is not consistent with the
+        state. This is called when the system of the ThermodynamicState
+        is set.
+
+        Parameters
+        ----------
+        system : simtk.openmm.System
+            The system to test.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def apply_to_context(self, context):
+        """Apply changes to the context to be consistent with the state."""
+        pass
+
+    @classmethod
+    @abc.abstractmethod
+    def standardize_system(cls, system):
+        """Standardize the given system.
+
+        ThermodynamicState relies on this method to create a standard
+        system that defines compatibility with another state or context.
+        The definition of a standard system is tied to the implementation
+        of apply_to_context. For example, if apply_to_context sets a
+        global parameter of the context, standardize_system should
+        set the default value of the parameter in the system to a
+        standard value.
+
+        Parameters
+        ----------
+        system : simtk.openmm.System
+            The system to standardize.
+
+        Raises
+        ------
+        ValueError
+            If the system cannot be standardized.
+
+        """
+        pass
+
+
+class CompoundThermodynamicState(ThermodynamicState):
+    """Thermodynamic state composed by multiple states.
+
+    Allows to extend a ThermodynamicState through composition rather
+    than inheritance.
+
+    The class dynamically inherits from the ThermodynamicState object
+    given in the constructor, and it preserves direct access to all its
+    methods and attributes. It is compatible also with subclasses of
+    ThermodynamicState, but it does not support objects which make use
+    of __slots__.
+
+    It is the user's responsibility to check that IComposableStates are
+    compatible to each other (i.e. that they do not depend on and/or
+    modify the same properties of the system). If this is not the case,
+    consider merging them into a single IComposableStates. If an
+    IComposableState needs to access properties of ThermodynamicState
+    (e.g. temperature, pressure) consider extending it through normal
+    inheritance.
+
+    It is not necessary to explicitly inherit from IComposableState for
+    compatibility as long as all abstract methods are implemented. All
+    its attributes and methods will still be directly accessible unless
+    they are masked by the main ThermodynamicState or by a IComposableState
+    that appeared before in the constructor argument composable_states.
+
+    After construction, changing the original thermodynamic_state or
+    any of the composable_states changes the state of the compound state.
+
+    Parameters
+    ----------
+    thermodynamic_state : ThermodynamicState
+        The main ThermodynamicState which holds the OpenMM 
```

Comment:
Do we have a defined behavior for when multiple states set the same thing? Or do we get ambiguous behavior?